core/src/main/java/com/linecorp/armeria/server/FallbackService.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.ResponseHeaders;
+
+/**
+ * Handles the {@link HttpRequest}s that are not matched by any user-specified {@link Route}s.
+ */
+final class FallbackService implements HttpService {
+
+    static final FallbackService INSTANCE = new FallbackService();
+
+    private FallbackService() {}
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        final RoutingContext routingCtx = ctx.routingContext();
+        final HttpStatusException cause = getStatusException(routingCtx);
+        if (cause.httpStatus() != HttpStatus.NOT_FOUND) {
+            throw cause;
+        }
+
+        // Handle 404 Not Found.
+        final String pathWithoutQuery = routingCtx.path();
+        if (pathWithoutQuery.charAt(pathWithoutQuery.length() - 1) == '/') {
+            // No need to send a redirect response because the request path already ends with '/'.
+            throw cause;
+        }
+
+        // Handle the case where '/path' (or '/path?query') doesn't exist
+        // but '/path/' (or '/path/?query') exists.
+        final String pathWithSlashAndWithoutQuery = pathWithoutQuery + '/';
+        if (!ctx.virtualHost()
+                .findServiceConfig(routingCtx.overridePath(pathWithSlashAndWithoutQuery))
+                .isPresent()) {
+            // No need to send a redirect response because '/path/' (or '/path/?query') does not exist.
+            throw cause;
+        }
+
+        // '/path/' (or '/path/?query') exists. Send a redirect response.
+        final String location;
+        final String pathWithQuery = req.path();
```

Comment:
Possibly more readable using `req.query()` since that's what we're concerned with

```java
if (req.query() == null) {
  location = req.path();
} else {
  location = req.path() + '?' + req.query();
}
```

I guess it's a tiny bit faster than current code (unless changed to `StringBuilder.append(CharSequence, to, end)`) too without intermediary `substring`

Reply:
Does this look better? https://github.com/line/armeria/pull/2255/commits/8b315fab0c5a7c88eb68f0281139ed5538a50001